# Exercise 4.8 - Task 

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Setting up the default path

path = r'C:\Users\peter\Desktop\Career Foundry - Data Analyst\Data Immersion\Achievement 4 - Python\01. Instacart Basket Analysis - June 2023'

In [3]:
# Importing the orders_products_ pkl file

df_orders_products_merged2 = pd.read_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'orders_products_busiest_days.pkl'))

#### Question 2: Find the aggregated mean, grouped by department_id, for the entire dataframe

In [4]:
# Using a single aggregation function to find the mean order_number, grouped by department_id

df_orders_products_merged2.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### Question 3: Analyze the results

#### These results are much different because we now have all of the department_id's available. The mean number of orders for each don't really vary too too much, but department_id 5 has the lowest mean at 15.2 and department_id 21 has the highest mean at 22.9.

#### Question 4: Create a loyalty flag for existing customers based on their number of orders. 

In [5]:
# First, we need to create a column called 'max_order' to see the highest number of orders each customer has made.

df_orders_products_merged2['max_order'] = df_orders_products_merged2.groupby(['user_id'])['order_number'].transform(np.max)

In [6]:
# Checking our results 

df_orders_products_merged2

order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32404854   1320836   202557    prior            17          2   
32404855     31526   202557    prior            18          5   
32404856    758936   203436    prior             1          2   
32404857   2745165   203436    prior             2          3   
32404858   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge    price_range_loc  \
0               77              7     9.0   both  Mid-range product   
1               77              7     9.0   both  Mid-range product   
2               77              7     9.0   both  Mid-range product   
3               77              7     9.0   both  Mid-range product   
4               77              7     9.0   both  Mid-range product   
...            ...            ...     ...    ...                ...   
32404854        64              7     3.7   both  Low-range product   
32404855        64              7     3.7   both  Low-range product   
32404856        50             19     6.9   both  Mid-range product   
32404857        50             19     6.9   both  Mid-range product   
32404858       130             14    10.3   both  Mid-range product   

             busiest_day   daily_activity busiest_period_of_day  max_order  
0         Regularly busy   Regularly busy        Average orders         10  
1         Regularly busy  Least busy days        Average orders         10  
2         Regularly busy  Least busy days        Average orders         10  
3             Least busy  Least busy days        Average orders         10  
4             Least busy  Least busy days           Most orders         10  
...                  ...              ...                   ...        ...  
32404854  Regularly busy   Regularly busy           Most orders         31  
32404855  Regularly busy   Regularly busy           Most orders         31  
3

In [7]:
# Using the loc() function to find values in our dataframe, in the 'max_order' column with certain criteria.
# From there, we are creating a new column called 'loyalty_flag' and assigning each a string value.

df_orders_products_merged2.loc[df_orders_products_merged2['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [8]:
df_orders_products_merged2.loc[(df_orders_products_merged2['max_order'] <= 40) & (df_orders_products_merged2['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [9]:
df_orders_products_merged2.loc[df_orders_products_merged2['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# Checking out new columns to see the results.

df_orders_products_merged2[['user_id', 'max_order', 'loyalty_flag']]

user_id  max_order      loyalty_flag
0               1         10      New customer
1               1         10      New customer
2               1         10      New customer
3               1         10      New customer
4               1         10      New customer
...           ...        ...               ...
32404854   202557         31  Regular customer
32404855   202557         31  Regular customer
32404856   203436          3      New customer
32404857   203436          3      New customer
32404858   205420         16  Regular customer

[32404859 rows x 3 columns]

In [ ]:
df_orders_products_merged2

#### Question 5: The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. 

In [11]:
# Running the agg() function with multiple statistics on the 'prices' column. We want to group by
# loyalty_flag so that we can compare the differences between the 3 groups.

df_orders_products_merged2.groupby('loyalty_flag').agg({'prices': ['mean', 'median', 'min', 'max']})

prices                     
                       mean median  min      max
loyalty_flag                                    
Loyal customer    10.386336    7.4  1.0  99999.0
New customer      13.294670    7.4  1.0  99999.0
Regular customer  12.495717    7.4  1.0  99999.0

#### Question 6: Create a spending flag for each user based on the average price across all their orders. Mean price < 10, flag them as "Low spender". Mean price >= 10, flag them as "High spender". 

In [12]:
# First, we need to create a new column called 'mean_spend' that is grouped by the user_id. This way, we can compare the
# average amount they spend across all their orders. 

df_orders_products_merged2['mean_spend'] = df_orders_products_merged2.groupby(['user_id'])['prices'].transform(np.mean)

In [13]:
# Checking our results

df_orders_products_merged2.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

  department_id  prices  _merge    price_range_loc     busiest_day  \
0             7     9.0    both  Mid-range product  Regularly busy   
1             7     9.0    both  Mid-range product  Regularly busy   
2             7     9.0    both  Mid-range product  Regularly busy   
3             7     9.0    both  Mid-range product      Least busy   
4             7     9.0    both  Mid-range product      Least busy   

    daily_activity busiest_period_of_day max_order  loyalty_flag  mean_spend  
0   Regularly busy        Average orders        10  New customer    6.367797  
1  Least busy days        Average orders        10  New customer    6.367797  
2  Least busy days        Average orders        10  New customer    6.367797  
3  Least busy days        Average orders        10  New customer    6.367797  
4  Least busy days           Most orders        10  New customer    6.367797  

[5 rows x 22 columns]

In [14]:
# The next step is to use the loc() function to run through each row and assign the named flags based on the new 
# mean_spend column we just created. 

df_orders_products_merged2.loc[df_orders_products_merged2['mean_spend'] < 10, 'spend_flag'] = 'Low spender'

In [15]:
df_orders_products_merged2.loc[df_orders_products_merged2['mean_spend'] >= 10, 'spend_flag'] = 'High spender'

In [16]:
# Checking our results with a value_counts() to see how it was distributed. 

df_orders_products_merged2['spend_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spend_flag, dtype: int64

#### Question 7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [17]:
# First, we need to create a new column called 'median_days_since' that is grouped by the user_id. This way, 
# we can compare the median number of days since their last order.

df_orders_products_merged2['median_days_since'] = df_orders_products_merged2.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [18]:
# Checking our results

df_orders_products_merged2.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

  _merge    price_range_loc     busiest_day   daily_activity  \
0   both  Mid-range product  Regularly busy   Regularly busy   
1   both  Mid-range product  Regularly busy  Least busy days   
2   both  Mid-range product  Regularly busy  Least busy days   
3   both  Mid-range product      Least busy  Least busy days   
4   both  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag mean_spend   spend_flag  \
0        Average orders        10  New customer   6.367797  Low spender   
1        Average orders        10  New customer   6.367797  Low spender   
2        Average orders        10  New customer   6.367797  Low spender   
3        Average orders        10  New customer   6.367797  Low spender   
4           Most orders        10  New customer   6.367797  Low spender   

   median_days_since  
0               20.5  
1               20.5  
2               20.5  
3               20.5  
4               20.5  

[5 rows x 24 columns]

In [19]:
# The next step is to use the loc() function to run through each row and assign the named flags based on the new 
# median_days_since column we just created. 

df_orders_products_merged2.loc[df_orders_products_merged2['median_days_since'] > 20, 'order_behavior_flag'] = 'Non-frequent customer'

In [20]:
df_orders_products_merged2.loc[(df_orders_products_merged2['median_days_since'] > 10) & (df_orders_products_merged2['median_days_since'] <= 20), 'order_behavior_flag'] = 'Regular customer'

In [21]:
df_orders_products_merged2.loc[df_orders_products_merged2['median_days_since'] <= 10, 'order_behavior_flag'] = 'Frequent customer'

In [22]:
# Checking our results with a value_counts() to see how it was distributed. 

df_orders_products_merged2['order_behavior_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_behavior_flag, dtype: int64

In [23]:
# Check for NaN values in the 'median_days_since' column. I wanted to investigate these rows a little more based on our
# value_counts() from above. 

nan_columns = df_orders_products_merged2[df_orders_products_merged2['median_days_since'].isnull()]

# Display the columns with NaN values

nan_columns

order_id  user_id eval_set  order_number  order_dow  \
13645692    895835   159838    prior             1          0   
17251990    895835   159838    prior             1          0   
17622767    895835   159838    prior             1          0   
24138593    895835   159838    prior             1          0   
25880002    895835   159838    prior             1          0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  ...    price_range_loc  busiest_day  \
13645692                  3          0  ...  Mid-range product  Busiest day   
17251990                  6          0  ...  Mid-range product  Busiest day   
17622767                  2          0  ...  Low-range product  Busiest day   
24138593                  5          0  ...  Mid-range product  Busiest day   
25880002                  1          0  ...  Low-range product  Busiest day   

          daily_activity  busiest_period_of_day max_order  loyalty_flag  \
13645692    Busiest days         Average orders         1  New customer   
17251990    Busiest days         Average orders         1  New customer   
17622767    Busiest days         Average orders         1  New customer   
24138593    Busiest days         Average orders         1  New customer   
25880002    Busiest days         Average orders         1  New customer   

         mean_spend   spend_flag median_days_since  order_behavior_flag  
13645692       7.42  Low spender               NaN                  NaN  
17251990       7.42  Low spender               NaN                  NaN  
17622767       7.42  Low spender               NaN                  NaN  
24138593       7.42  Low spender               NaN                  NaN  
25880002       7.42  Low spender               NaN                  NaN  

[5 rows x 25 columns]

In [24]:
# Exporting the new dataframe as a pkl file

df_orders_products_merged2.to_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'orders_products_grouped.pkl'))